### PR TITLE
Support host markdown in HostLocationProvider

### PR DESCRIPTION
**Pull request by jessesleeping**

### CODE REVIEW

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java [hunk 1]
```diff
@@ -148,7 +161,6 @@ public void close()
             catch (Exception e) {
                 log.warn("Error closing client", e);
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java]: Can you change the order of parameters? In airlift it's (e, message).

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java [hunk 2]
```diff
@@ -68,27 +75,33 @@ public RetryingPrestoThriftServiceProvider(@NonRetrying PrestoThriftServiceProvi
     @Override
     public PrestoThriftService anyHostClient()
     {
-        return new RetryingService(original::anyHostClient, retry);
+        return new RetryingService(() -> (DefaultPrestoThriftService) original.anyHostClient(), retry);
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java]: This code somewhat violates abstractions and creates tight coupling between components. In this case you're not working through the interface, but instead depend on another service giving you an object you need.
Try the following approach.
Extend `PrestoThriftService` interface to add `getLocation()` method to it. Then you can still use your wrappers, but they will be hidden from Default and Retrying providers and they will work though interfaces.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/DefaultPrestoThriftServiceProvider.java [hunk 3]
```diff
@@ -73,4 +89,73 @@ private PrestoThriftService connectTo(HostAddress host)
...
+
+    public static final class DefaultThriftService
+            implements PrestoThriftService
+    {
+        private final HostLocationHandle hostLocationHandle;
+        private final PrestoThriftService prestoThriftService;
+
+        public DefaultThriftService(PrestoThriftService prestoThriftService, HostLocationHandle hostLocationHandle)
+        {
+            requireNonNull(hostLocationHandle, "hostLocationHandle is null");
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/DefaultPrestoThriftServiceProvider.java]: inline with assignments

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java [hunk 4]
```diff
@@ -172,11 +173,8 @@ public RetryDriver withClassifier(Function<Exception, Exception> classifier)
         }
 
         return Futures.catchingAsync(resultFuture, Exception.class, e -> {
-            if (e instanceof InterruptedException) {
-                Thread.currentThread().interrupt();
-                throw propagate(e);
-            }
-            else if (retryStatus.shouldStopRetry(e)) {
+            onFailure.ifPresent(exceptionConsumer -> exceptionConsumer.accept(e));
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: Why did you remove propagation of InterruptedException?

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java [hunk 5]
```diff
@@ -162,7 +174,6 @@ public void close()
             catch (Exception e) {
                 log.warn(e, "Error closing client");
             }
-            client = null;
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java]: Don't remove this line, it's critical! Add a test to verify that we create a new client.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/location/StaticLocationProvider.java [hunk 6]
```diff
@@ -46,20 +48,68 @@ public StaticLocationProvider(StaticLocationConfig config)
...
+
+    private static class StaticHostLocationHandle
+            implements HostLocationHandle
+    {
+        private final HostAddress hostAddress;
+
+        public StaticHostLocationHandle(HostAddress hostAddress)
+        {
+            requireNonNull(hostAddress, "hostAddress is null");
+            this.hostAddress = hostAddress;
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/location/StaticLocationProvider.java]: inline assignment and check. Do it in all places where possible

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRestryingPrestoThriftServiceProvider.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        MockLocationAwarePrestoThriftServiceProvider mockServiceProvider = new MockLocationAwarePrestoThriftServiceProvider(TEST_HOST_LOCATION_PROVIDER);
+        ThriftConnectorConfig config = new ThriftConnectorConfig()
+                .setMaxRetryAttempts(3)
+                .setMinRetrySleepTime(new Duration(10, MILLISECONDS))
+                .setMaxRetrySleepTime(new Duration(20, MILLISECONDS));
+
+        RetryingPrestoThriftServiceProvider retryingServiceProvider =
+                new RetryingPrestoThriftServiceProvider(mockServiceProvider, retryExecutor, TEST_HOST_LOCATION_PROVIDER, config);
+
+        mockServiceProvider.setPendingFailuresToGenerate(2);
```
[the remark this answers is not in the document]
Comment (jessesleeping) [reply presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRestryingPrestoThriftServiceProvider.java]: Because the `MockLocationAwarePrestoThriftServiceProvider` is not stateless. It has configurable failures to be generated. We should have a provider per unit test.

[the remark this answers is not in the document]
Comment (AlekseiS) [reply presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRestryingPrestoThriftServiceProvider.java]: The question was about using a setter vs passing it in constructor.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftErrorCode.java [hunk 8]
```diff
@@ -24,7 +24,8 @@
 {
     THRIFT_SERVICE_CONNECTION_ERROR(1, EXTERNAL),
     THRIFT_SERVICE_INVALID_RESPONSE(2, EXTERNAL),
-    THRIFT_SERVICE_GENERIC_REMOTE_ERROR(3, EXTERNAL);
+    THRIFT_SERVICE_NO_AVAILABLE_HOSTS(3, EXTERNAL),
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftErrorCode.java]: You need to add new codes at the end and keep values of existing ones.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/DefaultPrestoThriftServiceProvider.java [hunk 9]
```diff
@@ -73,4 +89,71 @@ private PrestoThriftService connectTo(HostAddress host)
             throw new PrestoException(THRIFT_SERVICE_CONNECTION_ERROR, "Cannot connect to thrift host at " + host, e);
         }
     }
+
+    private static final class DefaultThriftService
+            implements ConnectedThriftService
+    {
+        private final HostLocationHandle hostLocationHandle;
+        private final PrestoThriftService prestoThriftService;
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/DefaultPrestoThriftServiceProvider.java]: Can you change the order: first service, then location?

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/DefaultPrestoThriftServiceProvider.java [hunk 10]
```diff
@@ -73,4 +89,71 @@ private PrestoThriftService connectTo(HostAddress host)
...
+
+        public DefaultThriftService(PrestoThriftService prestoThriftService, HostLocationHandle hostLocationHandle)
+        {
+            this.hostLocationHandle = requireNonNull(hostLocationHandle, "hostLocationHandle is null");
+            this.prestoThriftService = requireNonNull(prestoThriftService, "prestoThriftService is null");
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/DefaultPrestoThriftServiceProvider.java]: nit: I'd initialize in the order of parameters - first service, then location.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/DefaultPrestoThriftServiceProvider.java [hunk 11]
```diff
@@ -73,4 +89,71 @@ private PrestoThriftService connectTo(HostAddress host)
...
+        }
+
+        @Override
+        public HostLocationHandle getHostLocationHandle()
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/DefaultPrestoThriftServiceProvider.java]: move this method to the end of overriden methods

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java [hunk 12]
```diff
@@ -48,17 +47,17 @@
...
+            ConnectedThriftServiceProvider original,
+            @ForRetryDriver ListeningScheduledExecutorService retryExecutor,
+            ThriftConnectorConfig config)
     {
         this.original = requireNonNull(original, "original is null");
-        requireNonNull(retryExecutor, "retryExecutor is null");
-        requireNonNull(config, "config is null");
-
-        retry = RetryDriver.retry(retryExecutor)
+        retry = RetryDriver.retry(requireNonNull(retryExecutor, "retryExecutor is null"))
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java]: Why did you remove null check for config? Also, I'd keep retryExecutor null check separately.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java [hunk 13]
```diff
@@ -176,5 +158,30 @@ public void close()
...
+
+        private void processFailure(Exception e)
+        {
+            if (client != null) {
+                client.getHostLocationHandle().reportFailure(e);
+                this.close();
+            }
+        }
+
+        private PrestoThriftService getClient()
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java]: `getClient()` is a critical piece in implementing this interface, so it was previously in front of all methods.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java [hunk 14]
```diff
@@ -176,5 +158,30 @@ public void close()
...
+                return client;
+            }
+            client = clientSupplier.get();
+            return client;
+        }
+    }
+
+    private static Exception classifyException(Exception e)
+    {
+        if (e instanceof TApplicationException) {
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java]: You likely need to rebase to add TRuntime Exception

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/MockConnectedThriftServiceProvider.java [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+import com.facebook.presto.spi.HostAddress;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.facebook.presto.spi.HostAddress.fromParts;
+
+public class MockConnectedThriftServiceProvider
```
Comment (AlekseiS) [inline presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/MockConnectedThriftServiceProvider.java]: Don't call it mock. It's a test implementation. The convention in Presto is to prefix them with "Testing..". See examples in the code.

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/MockConnectedThriftServiceProvider.java [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+        implements ConnectedThriftServiceProvider
+{
+    private static final PrestoThriftServiceException RETRIABLE_EXCEPTION = new PrestoThriftServiceException("Mocking a retryable thrift failure", true);
```
Comment (AlekseiS) [inline presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/MockConnectedThriftServiceProvider.java]: Isn't it "retrYable"?

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/MockConnectedThriftServiceProvider.java [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+        {
+            failIfConfigured();
+            return ImmutableList.of();
+        }
+
+        @Override
+        public PrestoThriftNullableTableMetadata getTableMetadata(PrestoThriftSchemaTableName schemaTableName)
+                throws PrestoThriftServiceException
+        {
+            throw new UnsupportedOperationException("getTableMetadata is not supported in MockConnectedThriftService");
```
Comment (AlekseiS) [inline presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/MockConnectedThriftServiceProvider.java]: You don't need a message like this in test implementations

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/MockConnectedThriftServiceProvider.java [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    @Override
+    public ConnectedThriftService selectedHostClient(List<HostAddress> hosts)
+    {
+        openServiceCount.getAndIncrement();
+        return new MockConnectedThriftService(pendingFailuresToGenerate.getAndDecrement() > 0);
+    }
+
+    private final class MockConnectedThriftService
```
Comment (AlekseiS) [inline presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/MockConnectedThriftServiceProvider.java]: static?

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRetryingPrestoThriftServiceProvider.java [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+{
+    private ListeningScheduledExecutorService retryExecutor;
+
+    @BeforeClass
+    void init()
+    {
+        retryExecutor = listeningDecorator(newScheduledThreadPool(2, threadsNamed("test-thrift-retry-driver-%s")));
+    }
+
+    @AfterClass
```
Comment (AlekseiS) [inline presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRetryingPrestoThriftServiceProvider.java]: (alwaysRun=true)

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRetryingPrestoThriftServiceProvider.java [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+    void shutdown()
+    {
+        try {
+            retryExecutor.shutdownNow();
```
Comment (AlekseiS) [inline presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRetryingPrestoThriftServiceProvider.java]: if(retryExecutor!=null)

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRetryingPrestoThriftServiceProvider.java [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+        RetryingPrestoThriftServiceProvider retryingServiceProvider =
+                new RetryingPrestoThriftServiceProvider(mockServiceProvider, retryExecutor, config);
+
+        mockServiceProvider.setPendingFailuresToGenerate(2);
+
+        retryingServiceProvider.runOnAnyHost(PrestoThriftService::listSchemaNames);
+
+        assertEquals(mockServiceProvider.getOpenServiceCount(), 0);
+    }
+}
```
Comment (AlekseiS) [inline presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRetryingPrestoThriftServiceProvider.java]: since your test is small and your testing ("mock") implementation is very specific you might want to make it internal to this class.

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRetryingPrestoThriftServiceProvider.java [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+        MockConnectedThriftServiceProvider mockServiceProvider = new MockConnectedThriftServiceProvider();
+        ThriftConnectorConfig config = new ThriftConnectorConfig()
+                .setMaxRetryAttempts(3)
+                .setMinRetrySleepTime(new Duration(10, MILLISECONDS))
+                .setMaxRetrySleepTime(new Duration(20, MILLISECONDS));
+
+        RetryingPrestoThriftServiceProvider retryingServiceProvider =
+                new RetryingPrestoThriftServiceProvider(mockServiceProvider, retryExecutor, config);
+
+        mockServiceProvider.setPendingFailuresToGenerate(2);
```
Comment (AlekseiS) [inline presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRetryingPrestoThriftServiceProvider.java]: Why not move pending failuresToGenerate initialization to constructor?

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRetryingPrestoThriftServiceProvider.java [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+
+        retryingServiceProvider.runOnAnyHost(PrestoThriftService::listSchemaNames);
+
+        assertEquals(mockServiceProvider.getOpenServiceCount(), 0);
```
Comment (AlekseiS) [inline presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRetryingPrestoThriftServiceProvider.java]: It doesn't seem to be verifying the case with removed `null` that you had before. Did it fail in that case?
If that's what you testing it might be good to count the number of times client was requested explicitly.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/location/StaticLocationProvider.java [hunk 24]
```diff
@@ -42,24 +45,77 @@ public StaticLocationProvider(StaticLocationConfig config)
         Collections.shuffle(this.hosts);
     }
 
+    @VisibleForTesting
+    public StaticLocationProvider(List<HostAddress> hosts)
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/location/StaticLocationProvider.java]: this constructor is not used

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/MockConnectedThriftServiceProvider.java (hunk 16 above)
Comment (jessesleeping) [reply presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/MockConnectedThriftServiceProvider.java]: :p

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/MockConnectedThriftServiceProvider.java (hunk 18 above)
Comment (jessesleeping) [reply presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/MockConnectedThriftServiceProvider.java]: It needs to update the `openServiceCount` when closing the thrift service.

Comment (jessesleeping) [reply presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/MockConnectedThriftServiceProvider.java]: I removed the `failedService` member from this class and directly relied on the `pendingFailuresToGenerate` in the outer class.

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/MockConnectedThriftServiceProvider.java (hunk 17 above)
Comment (jessesleeping) [reply presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/MockConnectedThriftServiceProvider.java]: Need to throw something and have some words in the message. I think it should be fine.

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRetryingPrestoThriftServiceProvider.java (hunk 23 above)
Comment (jessesleeping) [reply presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRetryingPrestoThriftServiceProvider.java]: It did fail in that case. I added check to explicitly check the number of open calls and close calls.

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/MockConnectedThriftServiceProvider.java (hunk 17 above)
Comment (AlekseiS) [reply presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/MockConnectedThriftServiceProvider.java]: You can keep the text, but remove class and method names from the message.

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRetryingPrestoThriftServiceProvider.java [hunk 25]
```diff
@@ -0,0 +1,208 @@
...
+        assertEquals(testingServiceProvider.getOpenServiceCount(), 3, "Must open thrift connections for 3 times");
+        assertEquals(testingServiceProvider.getCloseServiceCount(), 3, "Must close all 3 opened connections");
+    }
+
+    private static class TestingConnectedThriftServiceProvider
+            implements ConnectedThriftServiceProvider
+    {
+        private static final PrestoThriftServiceException RETRYABLE_EXCEPTION = new PrestoThriftServiceException("Mocking a retryable thrift failure", true);
+        private static final HostAddress DUMMY_HOST_ADDRESS = fromParts("test_host", 1234);
+        private static final HostLocationHandle DUMMY_HOST_LOCATION_HANDLE = new HostLocationHandle()
```
Comment (AlekseiS) [inline presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRetryingPrestoThriftServiceProvider.java]: You would be able to use `SimpleHostLocationHandle` here as well

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRetryingPrestoThriftServiceProvider.java [hunk 26]
```diff
@@ -0,0 +1,208 @@
...
+                retryExecutor.shutdownNow();
+            }
+        }
+        catch (Exception ignore) {
+        }
+    }
+
+    @Test
+    void testRetryService()
+            throws Exception
```
Comment (AlekseiS) [inline presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRetryingPrestoThriftServiceProvider.java]: IDE tells me that you don't throw checked exceptions in this method

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRetryingPrestoThriftServiceProvider.java [hunk 27]
```diff
@@ -0,0 +1,198 @@
...
+
+        assertEquals(testingServiceProvider.getOpenServiceCount(), 3, "Must open thrift connections for 3 times");
+        assertEquals(testingServiceProvider.getCloseServiceCount(), 3, "Must close all 3 opened connections");
+    }
+
+    private static class TestingConnectedThriftServiceProvider
+            implements ConnectedThriftServiceProvider
+    {
+        private static final PrestoThriftServiceException RETRYABLE_EXCEPTION = new PrestoThriftServiceException("Mocking a retryable thrift failure", true);
+        private static final HostAddress DUMMY_HOST_ADDRESS = fromParts("test_host", 1234);
```
Comment (AlekseiS) [inline presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRetryingPrestoThriftServiceProvider.java]: you can inline it
